### PR TITLE
handle nested generics in Generics::type_param/region_param

### LINKED ISSUE #44952: ICE: unexpected panic in librustc/ty/mod.rs
## Source
[lib.rs](https://gist.github.com/joshlf/d5ebb370ccdfc8a8af39923b5a788970)

## Compiler output

```text
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.0-nightly (6f87d20a7 2017-09-29) running on x86_64-apple-darwin

thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `0`', src/librustc/ty/mod.rs:751:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #44952.

r? @eddyb

### CODE REVIEW

File: src/librustc/ty/mod.rs [hunk 1]
```diff
@@ -742,14 +742,28 @@ impl Generics {
...
     }
 
-    pub fn type_param(&self, param: &ParamTy) -> &TypeParameterDef {
-        assert_eq!(self.parent_count(), 0);
-        &self.types[param.idx as usize - self.has_self as usize - self.regions.len()]
+    pub fn type_param(&'tcx self,
+                      param: &ParamTy,
+                      tcx: TyCtxt<'a, 'gcx, 'tcx>) -> &TypeParameterDef {
+        if let Some(idx) = param.idx.checked_sub(self.parent_count() as u32) {
+            &self.types[idx as usize - self.has_self as usize - self.regions.len()]
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: This doesn't seem to take into account the fact that `Self` is at index `0` *before* lifetimes.

**Comment by eddyb** [comment]
r=me w/ `Self` fixed (although I believe the current uses don't care)

File: src/librustc/ty/mod.rs (hunk 1 above)
Comment (arielb1) [reply src/librustc/ty/mod.rs]: How are you supposed to detect `Self`?

Comment (eddyb) [reply src/librustc/ty/mod.rs]: `has_self` and the index is `0`.

**Comment by arielb1** [comment]
@eddyb 

done

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -742,14 +742,36 @@ impl Generics {
...
+    pub fn type_param(&'tcx self,
+                      param: &ParamTy,
+                      tcx: TyCtxt<'a, 'gcx, 'tcx>)
+                      -> Option<&TypeParameterDef> {
+        if let Some(idx) = param.idx.checked_sub(self.parent_count() as u32) {
+            let type_param_start = (self.has_self as usize) + self.regions.len();
+            if let Some(idx) = (idx as usize).checked_sub(type_param_start) {
+                Some(&self.types[idx])
+            } else {
+                None
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: Why not `assert_eq!((idx, self.has_self), (0, true));` followed by `&self.types[0]`?

Comment (arielb1) [reply src/librustc/ty/mod.rs]: `idx` of 1 is also mapped to `1 - 1 = 0`, so I don't think `&self.types[0]` is correct here and `Self` doesn't have a `TypeParameterDef`.

**Pull request merged**